### ISSUE TITLE
implement PAMI transport

### ISSUE THREAD

**Issue by sebhtml** [GeneAssembly/biosal#453]
http://code.google.com/p/pami-examples/

code will go in:

[boisvert@cetuslac1 biosal]$ find engine/thorium/transport/pami
engine/thorium/transport/pami
engine/thorium/transport/pami/pami_active_buffer.c
engine/thorium/transport/pami/pami_active_buffer.h
engine/thorium/transport/pami/pami_transport.c
engine/thorium/transport/pami/pami_transport.h

**Comment by sebhtml** [comment]
use this example:

http://code.google.com/p/pami-examples/source/browse/trunk/function/send.c

**Comment by sebhtml** [comment]
in charm++:

http://charm.cs.uiuc.edu/gerrit/#/q/status:open,n,z

http://charm.cs.uiuc.edu/gerrit/gitweb?p=charm.git;a=summary

MPI abstract machine:

http://charm.cs.uiuc.edu/gerrit/gitweb?p=charm.git;a=tree;f=src/arch/mpi;h=5406efc67e55396b88cbc24d1dc060bca46ca2e8;hb=HEAD

pami:

http://charm.cs.uiuc.edu/gerrit/gitweb?p=charm.git;a=tree;f=src/arch/pami;h=9a546878d6b7d4b4b6372a530efb0ab134a345d2;hb=HEAD

**Comment by sebhtml** [comment]
in mpi machine in Converse:

init: LrtsInit
destroy: CkDieNow
send: MPISendOneMsg
receive: PumpMsgs

in pami machine:

send: machine_send calls PAMI_Send

**Comment by sebhtml** [comment]
PAMI documentation from IBM:

https://github.com/jeffhammond/HPCInfo/wiki/PAMI

Doxygen:

https://bgq1.epfl.ch/navigator/resources/doc/pami/examples.html

IBM book:

http://publibfp.dhe.ibm.com/epubs/pdf/a2322733.pdf

**Comment by sebhtml** [comment]
https://www.alcf.anl.gov/files/Mysteries_of_the-Deep_What_happens_inside_MPI_on_Blue%20GeneQ_why_it_matters_0.pdf

**Comment by sebhtml** [comment]
TRANSPORT rank: 206 size: 512 implementation: PAMI

**Comment by sebhtml** [comment]
Huy will do it.

**Comment by sebhtml** [comment]
Design note from Huy:

Here is step-by-step guide to program on BG/Q using PAMI. I am writing
this to give you an introduction of how PAMI works. I will write PAMI
code, but this would make it easier for all of us.

General initialization:
1. Create a client: PAMI_Client_create. You can get rank, size, system
clock cycles here. You will need a client's name.
2. Create a context: PAMI_Context_createv
3. Set a dispatch: PAMI_Dispatch_set: to register an dispatch with a
callback function. 3 inputs: dispatch ID, callback function, session
variable. PAMI will attach the dispatch ID with sending message. When a
new message come, PAMI will check the dispatch ID of incoming message.
It will give the message to the corresponding callback function f. The
session variable is used to communicate between main program and
callback function.

Memory related:
4. Malloc a memory or get it from a pool (if you manage memory on your
own style in GeneAssembly)
5. Create memory regions: PAMI_Memregion_create
6. Broadcast the memory regions to other endpoints (whoever going to
need it later). Usually use PAMI_Send_immediate to broadcast memory regions.
1. Create endpoint: For each destination, you need to create an
   endpoint: PAMI_Endpoint_create.
2. Send-Wait-Recv: To send you can use either: PAMI_Send_immediate,
   PAMI_Send, or to put PAMI_Put or to get PAMI_Get. They are all
   non-blocking communication.

On the sending side: if PAMI_Send_immediate is used, CPU will be
involved into communication until it's done. In other cases, the
send/put/get request is put in a queue. You need to check the queue to
make sure that the request is done (to reuse the memory for example).
You need to call PAMI_Context_advance() so PAMI will check the queue for
you.

To receive, you need to call: PAMI_Context_advance() so the PAMI engine
can check the queue to see if any messages are coming. As soon as a
message comes, PAMI will match its dispatch ID with a registered
callback function. The function will then be called to process the message.

You can also specify a data function (not callback function) to process
at packet level (512 bytes), when a packet come, the data function can
be called to process the packet. But the function should be light-weight
such as copy, sum, subtract...

We can discuss more details if you want. I'm a bit busy to start coding,
but discussion is open.

Huy.

**Comment by sebhtml** [comment]
I am meeting with Huy tomorrow.

**Comment by sebhtml** [comment]
Bug:
 /gpfs/mira-fs1/projects/CompBIO-data/debugging/test-spate-627-512x16-13.job

For starters, here is the stack:

[boisvert@cetuslac1 test-spate-627-512x16-13.job]$ addr2line -e spate < test-spate-627-512x16-13.output-stack.txt
/home/abui/biosal/core/system/tracer.c:40
/home/abui/biosal/core/system/memory.c:91
/home/abui/biosal/core/system/memory_pool.c:194
/home/abui/biosal/core/system/memory_pool.c:162
/home/abui/biosal/core/structures/vector.c:247
/home/abui/biosal/engine/thorium/modules/binomial_tree_message.c:46
/home/abui/biosal/engine/thorium/actor.c:785
/home/abui/biosal/engine/thorium/actor.c:941
/home/abui/biosal/engine/thorium/actor.c:1812
/home/abui/biosal/engine/thorium/worker.c:1238
/home/abui/biosal/engine/thorium/worker.c:1139
/home/abui/biosal/engine/thorium/worker.c:319
/bgsys/drivers/V1R2M2/ppc64/toolchain/gnu/glibc-2.12.2/nptl/pthread_create.c:322
??:0

So basically, the message contains an invalid value when the vector is unpacking:

/home/abui/biosal/core/structures/vector.c:247
            self->data = bsal_memory_pool_allocate(self->memory, self->maximum_size \* self->element_size);

> Weirdly enough. My transport lib passed all tests.
> process/1000000 PASSED 100197/100197, FAILED 0/100197
> process/1000001 PASSED 99463/99463, FAILED 0/99463
> process/1000002 PASSED 100381/100381, FAILED 0/100381
> process/1000003 PASSED 99959/99959, FAILED 0/99959

**Comment by sebhtml** [comment]
It is failing with action ACTION_BINOMIAL_TREE_SEND, which can send fairly large messages.

**Comment by huyba** [comment]
MAXIMUM size of 512KB has been tested and passed (all testcases)
process/1000000 PASSED 100197/100197, FAILED 0/100197
process/1000001 PASSED 99463/99463, FAILED 0/99463
process/1000002 PASSED 100381/100381, FAILED 0/100381
process/1000003 PASSED 99959/99959, FAILED 0/99959

MAXIMUM size of 2MB has been tested and passed (all testcases)
process/1000000 PASSED 100290/100290, FAILED 0/100290
process/1000002 PASSED 100101/100101, FAILED 0/100101
process/1000001 PASSED 100057/100057, FAILED 0/100057
process/1000003 PASSED 99552/99552, FAILED 0/99552

Is testing with maximum size of 4MB.

**Comment by sebhtml** [comment]
OK. So the next step I guess would be to send more messages concurrently in the process actor.

**Comment by huyba** [comment]
All testcases have passed for the test of 512 nodes, 16 threads/node, maximum data size of 2MB. What are we supposed to do next now????

**Comment by sebhtml** [comment]
You could commit your code and put this information in the commit message.
I'll review it and merge it into master.

Does it work on Iowa Continuous Corn ?

Otherwise, we need to test with more concurrent messages.

Also, the -threads-per-node option for the transport app does not do anything because
the initial actors are not spawning anything.

I can modify the test app so that it sens messages concurrently.

**Comment by huyba** [comment]
No, it did not work on Iowa Continuous Conrn.
We need to test with more concurrent messages. Please add more concurrent threads to see how it behaves.

**Comment by sebhtml** [comment]
Did you make changes to the app ?

**Comment by huyba** [comment]
I did not change any code of the app, jus the pami transport, but not much either and they did not work correct now. So I will commit later.

**Comment by huyba** [comment]
For now, can you change the parameter in the transport_tester so I can run more test?

**Comment by sebhtml** [comment]
I merged your code.

There are a couple of issues regarding the coding style:

Detected issues: 24
engine/thorium/transport/pami/pami_transport.c:45:    for(i = 0; i < NUM_RECV_BUFFERS; i++) {
engine/thorium/transport/pami/pami_transport.c:116:    for(i = 0; i < NUM_RECV_BUFFERS; i++) {
engine/thorium/transport/pami/pami_transport.c:155:    if(nbytes <= MAX_SHORT_MESSAGE_LENGTH) {
engine/thorium/transport/pami/pami_transport.c:166:        if(result != PAMI_SUCCESS) {
engine/thorium/transport/pami/pami_transport.c:186:        if(result != PAMI_SUCCESS) {
engine/thorium/transport/pami/pami_transport.c:199:    if(bsal_fast_queue_size(pami_transport->recv_queue) > 0) {
engine/thorium/transport/pami/pami_transport.c:217:    if(bsal_fast_queue_size(pami_transport->send_queue) > 0) {
engine/thorium/transport/pami/pami_transport.c:265:    if(recv == NULL) {
engine/thorium/transport/pami/pami_transport.c:157:     parameter.dispatch        = RECV_MESSAGE_DISPATCH_ID;
engine/thorium/transport/pami/pami_transport.c:159:        parameter.header.iov_len  = 0;
engine/thorium/transport/pami/pami_transport.c:160:     parameter.data.iov_base   = (void *)buffer;
engine/thorium/transport/pami/pami_transport.c:161:        parameter.data.iov_len    = nbytes;
engine/thorium/transport/pami/pami_transport.c:178:     param_send.send.data.iov_base  = buffer;
engine/thorium/transport/pami/pami_transport.c:180:        param_send.events.cookie        = (void *)send_cookie;
engine/thorium/transport/pami/pami_transport.c:181:        param_send.events.local_fn      = send_done_fn;
engine/thorium/transport/pami/pami_transport.c:182:        param_send.events.remote_fn     = NULL;
engine/thorium/transport/pami/pami_transport.c:273:        recv->cookie   = (void *)recv_cookie;
engine/thorium/transport/pami/pami_transport.c:274:        recv->type     = PAMI_TYPE_BYTE;
engine/thorium/transport/pami/pami_transport.c:275:        recv->addr     = buffer;
engine/thorium/transport/pami/pami_transport.c:276:        recv->offset   = 0;
engine/thorium/transport/pami/pami_transport.c:277:        recv->data_fn  = PAMI_DATA_COPY;
engine/thorium/transport/pami/pami_transport.h:76:  
engine/thorium/transport/pami/pami_transport.c:39:  
engine/thorium/transport/pami/pami_transport.c:123:   

see Documentation/CodingStyle.md

Most of them are related to indentation or to a missing space after if or for.

Now, regarding your inquiry:
- I moved the performance apps in performance/
- make CONFIG_PAMI=y is required to build with PAMI support.

The app is in performance/transport_tester/transport_tester now (because it is not
a genomics app).

New options:
-maximum-buffer-size <maximum_buffer_size> (default: 524288, in bytes)
-event-count <event_count> (default: 100000, this is the number of send events for each actor)
-concurrent-event-count <value> (default: 8, this is the maximum number of concurrent send events per actor)

Example:

[boisvert@bigmem biosal]$ mpiexec -n 4 performance/transport_tester/transport_tester -threads-per-node 8 -maximum-buffer-size 10000 -event-count 200000 -concurrent-event-count 8| tee log

**Comment by huyba** [comment]
I have tested my code with 4M for 512 nodes and all testcases passed. I will test more with this code, modify my code as you commented and update repo.

**Comment by sebhtml** [comment]
OK.

Did you try with more concurrent events ?

**Comment by huyba** [comment]
Not yet. I'm trying now. I'll let you know the results.

**Comment by huyba** [comment]
./scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh used make CONFIG_PAMI=y, and I also used -transport pami_transport. But when I executed the transport_tester, it still printed out: DEBUG TRANSPORT -> mpi1_pt2pt_nonblocking_transport

Is that set to mpi?

**Comment by sebhtml** [comment]
You also need to add -DPAMI_CONFIG to your CFLAGS.

[seb@localhost biosal]$ git diff
diff --git a/scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh b/scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh
index f11049f..fd7dee6 100755
--- a/scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh
+++ b/scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh
@@ -12,7 +12,7 @@ CFLAGS="$CFLAGS -L/soft/libraries/alcf/current/xl/ZLIB/lib -I/soft/libraries/alc
 #exit

 make clean
-make -j 8 CFLAGS="$CFLAGS" CONFIG_PAMI=y \
+make -j 8 CFLAGS="$CFLAGS -DPAMI_CONFIG" CONFIG_PAMI=y \
     applications/argonnite_kmer_counter/argonnite \
     applications/spate_metagenome_assembler/spate \
     examples/example_ring

**Comment by sebhtml** [comment]
I fixed that, it should be fine with "make CONFIG_PAMI=y" now. The macro CONFIG_PAMI was not defined. 
https://github.com/sebhtml/biosal/commit/d14accb79613d4a7887badfba7e0f0ba7fd5a594

**Comment by huyba** [comment]
make -j 8 CFLAGS="$CFLAGS" CONFIG_PAMI=y \

#!/bin/bash

qsub \
--env PAMID_THREAD_MULTIPLE=1 \
-A Performance \
-n 4 \
-t 00:60:00 \
-O transport_tester-8 \
--mode c1 \
transport_tester \
-maximum-buffer-size 2097152 \
-event-count 100000 \
-concurrent-event-count 8 \
- transport pami_transport

But it still print out: DEBUG TRANSPORT -> mpi1_pt2pt_nonblocking_transpor

And it failed:

PROGRESS 29000/100000
DEBUG Error bsal_memory_allocate returned (nil), 2097152 bytes
used / total -> 11661406208 / 17163091968
bsal_tracer_print_stack_backtrace
Stack backtrace has 22 frames
#0  [0x103667c]
#1  [0x1035dc0]

I think it's used pami, but can you confirm? Can we know more detail of how the error happened. The test is transport_tester-8.\* under /gpfs/mira-fs0/projects/Performance/abui/microbm/biosal/tester/

**Comment by sebhtml** [comment]
It is using mpi1_pt2pt non blocking indeed because your script has an error:

```
[boisvert@cetuslac1 tester]$ tail transport_tester-8.sh 
-A Performance \
-n 4 \
-t 00:60:00 \
-O transport_tester-8 \
--mode c1 \
transport_tester \
-maximum-buffer-size 2097152 \
-event-count 100000 \
-concurrent-event-count 8 \
- transport pami_transport   <<<<<<<<<<<<<<<<<<< space between - and transport
```

From the stack, it seems that the memory is being exausted, yet the line
"used / total -> 11661406208 / 17163091968" says otherwise. This is strange.

Here is the stack:

[boisvert@cetuslac1 tester]$ addr2line -e transport_tester < ~/transport_tester-8.stack
/home/abui/workspace/biosal/core/system/tracer.c:40
/home/abui/workspace/biosal/core/system/memory.c:99
/home/abui/workspace/biosal/core/system/memory_pool.c:216
/home/abui/workspace/biosal/core/system/memory_pool.c:194
/home/abui/workspace/biosal/engine/thorium/worker.c:245
/home/abui/workspace/biosal/engine/thorium/actor.c:423
/home/abui/workspace/biosal/engine/thorium/modules/send_helpers.c:282
/home/abui/workspace/biosal/performance/transport_tester/process.c:225
/home/abui/workspace/biosal/engine/thorium/dispatcher.c:187
/home/abui/workspace/biosal/engine/thorium/actor.c:1238
/home/abui/workspace/biosal/performance/transport_tester/process.c:88
/home/abui/workspace/biosal/engine/thorium/actor.c:978
/home/abui/workspace/biosal/engine/thorium/actor.c:1812
/home/abui/workspace/biosal/engine/thorium/worker.c:1239
/home/abui/workspace/biosal/engine/thorium/worker.c:1140
/home/abui/workspace/biosal/engine/thorium/worker_pool.c:205
/home/abui/workspace/biosal/engine/thorium/node.c:1894
/home/abui/workspace/biosal/engine/thorium/node.c:745
/home/abui/workspace/biosal/engine/thorium/thorium_engine.c:30
/home/abui/workspace/biosal/performance/transport_tester/main.c:8
/bgsys/drivers/V1R2M2/ppc64/toolchain/gnu/glibc-2.12.2/csu/../csu/libc-start.c:226
/bgsys/drivers/V1R2M2/ppc64/toolchain/gnu/glibc-2.12.2/csu/../sysdeps/unix/sysv/linux/powerpc/libc-start.c:194
??:0

EDIT 2014-09-02: this memory leak / abstraction leak has been fixed in master:
https://github.com/sebhtml/biosal/commit/4f51c9cd8343be477f052991576f7bd5aa2323b4

**Comment by huyba** [comment]
I have tested pami option with 4nodes, 8 concurrent threads, data size 2MB maximum. All test cases passed.

process/1000016 PASSED 99821/99821, FAILED 0/99821
process/1000019 PASSED 99640/99640, FAILED 0/99640
process/1000018 PASSED 99558/99558, FAILED 0/99558
process/1000017 PASSED 99546/99546, FAILED 0/99546

I'm testing 512 nodes, 8 concurrent threads, data size of 2MB maximum for PAMI option. I'll posted result when I got it.

**Comment by huyba** [comment]
Tested with 512 nodes, 8 concurrent threads, data size of 2MB maximum, PAMI option. Lots of invalid data. For example:

TRANSPORT FAILED source: 1002320 (336) destination: 1002963 (467) tag: ACTION_PING count: 1433888 expected_checksum: 5352035122396698533 actual_checksum: 3711446915899317171

The test is transport_tester-11.\* in /gpfs/mira-fs0/projects/Performance/abui/microbm/biosal/tester/ 

Can you have a look and tell me more?

**Comment by sebhtml** [comment]
You can edit the app ( https://github.com/sebhtml/biosal/blob/master/performance/transport_tester/process.c#L110 ) to call bsal_debugger_examine on the buffer ( https://github.com/sebhtml/biosal/blob/master/core/system/debugger.h#L43 )

The expected buffer content should look like this:

0 1 2 3 4 5 ... 253 254 255 0 1 2 3 4 ...

The code that sets the content is here:
https://github.com/sebhtml/biosal/blob/master/performance/transport_tester/process.c#L212

**Comment by sebhtml** [comment]
Code review: https://github.com/sebhtml/biosal/pull/2

**Comment by sebhtml** [comment]
This implementation works:
https://github.com/GeneAssembly/biosal/tree/mirror/engine/thorium/transport/pami

**Comment by huyba** [comment]
Running time using PAMI vs. MPI:

MPI - 512 nodes
TIMER [Load input / Count input data] 1 minutes, 0.024452 seconds
TIMER [Load input / Distribute input data] 1 minutes, 13.344955 seconds
TIMER [Load input] 2 minutes, 13.369400 seconds
TIMER [Build assembly graph / Distribute vertices] 10 minutes, 38.107788 seconds
TIMER [Build assembly graph / Distribute arcs] 17 minutes, 41.421753 seconds
TIMER [Build assembly graph] 28 minutes, 19.529541 seconds
TIMER [Run actor computation] 30 minutes, 35.615479 seconds

PAMI - 512 nodes
TIMER [Load input / Count input data] 1 minutes, 1.068619 seconds
TIMER [Load input / Distribute input data] 1 minutes, 12.936531 seconds
TIMER [Load input] 2 minutes, 14.005157 seconds
TIMER [Build assembly graph / Distribute vertices] 10 minutes, 42.060852 seconds
TIMER [Run actor computation] 30 minutes, 25.466431 seconds
TIMER [Build assembly graph / Distribute arcs] 17 minutes, 26.833374 seconds
TIMER [Build assembly graph] 28 minutes, 8.894165 seconds

MPI - 1K nodes
TIMER [Load input / Count input data] 44.235592 seconds
TIMER [Load input / Distribute input data] 1 minutes, 44.567924 seconds
TIMER [Load input] 2 minutes, 28.803513 seconds
TIMER [Build assembly graph / Distribute vertices] 7 minutes, 47.384521 seconds
TIMER [Run actor computation] 20 minutes, 24.319946 seconds
TIMER [Build assembly graph / Distribute arcs] 10 minutes, 4.386292 seconds
TIMER [Build assembly graph] 17 minutes, 51.770874 seconds

PAMI - 1K nodes
TIMER [Load input / Count input data] 43.986809 seconds
TIMER [Load input / Distribute input data] 1 minutes, 43.696808 seconds
TIMER [Load input] 2 minutes, 27.683624 seconds
TIMER [Build assembly graph / Distribute vertices] 8 minutes, 38.307617 seconds
TIMER [Run actor computation] 22 minutes, 22.072021 seconds
TIMER [Build assembly graph / Distribute arcs] 11 minutes, 11.398987 seconds
TIMER [Build assembly graph] 19 minutes, 49.706665 seconds

MPI - 1536 nodes
TIMER [Load input / Count input data] 35.129623 seconds
TIMER [Load input / Distribute input data] 2 minutes, 16.061493 seconds
TIMER [Load input] 2 minutes, 51.191116 seconds
TIMER [Build assembly graph / Distribute vertices] 6 minutes, 35.233673 seconds
Seem to miss something here. I'll rerun the experiment.

PAMI - 1536 nodes
TIMER [Load input / Count input data] 34.771259 seconds
TIMER [Load input / Distribute input data] 2 minutes, 14.326385 seconds
TIMER [Load input] 2 minutes, 49.097656 seconds
TIMER [Build assembly graph / Distribute vertices] 6 minutes, 41.999481 seconds
TIMER [Build assembly graph / Distribute arcs] 8 minutes, 17.209442 seconds
TIMER [Build assembly graph] 14 minutes, 59.208923 seconds

MPI - 2K nodes
TIMER [Load input / Count input data] 36.146912 seconds
TIMER [Load input / Distribute input data] 2 minutes, 50.586319 seconds
TIMER [Load input] 3 minutes, 26.733231 seconds
TIMER [Build assembly graph / Distribute vertices] 6 minutes, 19.435547 seconds
TIMER [Build assembly graph / Distribute arcs] 7 minutes, 26.814911 seconds
TIMER [Build assembly graph] 13 minutes, 46.250427 seconds

PAMI - 2K nodes
TIMER [Load input / Count input data] 39.072330 seconds
TIMER [Load input / Distribute input data] 2 minutes, 48.061676 seconds
TIMER [Load input] 3 minutes, 27.134003 seconds
TIMER [Build assembly graph / Distribute vertices] 7 minutes, 50.768372 seconds
TIMER [Build assembly graph / Distribute arcs] 8 minutes, 27.794678 seconds
TIMER [Build assembly graph] 16 minutes, 18.563049 seconds
TIMER [Run actor computation] 19 minutes, 54.253784 seconds

**Comment by sebhtml** [comment]
13-minute 2048-node report for Q/A department
(template is from https://github.com/GeneAssembly/biosal/blob/mirror/Documentation/Quality_Assurance.md )

**JobName**
All files with 
test-spate-627-2048x16-1-MPI_nonblk 
name.

keywords: 13 minutes 2048 16 2048x16

**Machine**
cetus http://en.wikipedia.org/wiki/Cetus

## **AllocationStatus**

**Path**
/gpfs/mira-fs0/projects/Performance/abui/microbm/biosal/spate/

**Commit**
unknown

**Toolchain**
/soft/compilers/wrappers/gcc/mpicc

but with the script for XL (without using XL):

$ ls scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh 
scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh
$ cat scripts/IBM_Blue_Gene_Q/build-blue-gene-q.sh 
#!/bin/bash

 # http://www-01.ibm.com/support/docview.wss?uid=swg27022103&aid=1

 # -O5 to enable -ipa
 # -s to strip symbols

CFLAGS="-I. -g -O3 -qmaxmem=-1 -qarch=qp -qtune=qp -DBSAL_DEBUGGER_ENABLE_ASSERT "
CFLAGS="$CFLAGS -L/soft/libraries/alcf/current/xl/ZLIB/lib -I/soft/libraries/alcf/current/xl/ZLIB/include"

#echo $CFLAGS
#exit

make clean
make -j 8 CFLAGS="$CFLAGS" CONFIG_PAMI=n \
    applications/argonnite_kmer_counter/argonnite \
    applications/spate_metagenome_assembler/spate \
    examples/example_ring \
    performance/transport_tester/transport_tester

**Script**
[boisvert@cetuslac1 spate]$ cat test-spate-627-2048x16-1-MPI_nonblk.sh 
#!/bin/bash

qsub \
--env PAMID_THREAD_MULTIPLE=1 \
-A Performance \
-n 2048 \
-t 00:30:00 \
-O test-spate-627-2048x16-1-MPI_nonblk \
--mode c1 \
spate -print-load -threads-per-node 16 \
-transport mpi1_pt2pt_nonblocking_transport \
-k 43 Iowa_Continuous_Corn/*.fastq \
-o test-spate-627-2048x16-1-MPI_nonblk

**Submission**
not available.

[boisvert@cetuslac1 spate]$ ls -l test-spate-627-2048x16-1-MPI_nonblk -d 
drwxr-sr-x 2 abui Performance 12845056 Sep  8 02:56 test-spate-627-2048x16-1-MPI_nonblk
[boisvert@cetuslac1 spate]$ ls -l test-spate-627-2048x16-1-MPI_nonblk.\* 
-rw-r--r-- 1 abui Performance      7731 Sep  8 02:57 test-spate-627-2048x16-1-MPI_nonblk.cobaltlog
-rw-r--r-- 1 abui Performance      1986 Sep  8 02:57 test-spate-627-2048x16-1-MPI_nonblk.error
-rw-r--r-- 1 abui Performance 257547556 Sep  8 02:56 test-spate-627-2048x16-1-MPI_nonblk.output
-rwxr-xr-x 1 abui Performance       309 Sep  8 00:15 test-spate-627-2048x16-1-MPI_nonblk.sh
[boisvert@cetuslac1 spate]$ ls -l test-spate-627-2048x16-1-MPI_nonblk -d
drwxr-sr-x 2 abui Performance 12845056 Sep  8 02:56 test-spate-627-2048x16-1-MPI_nonblk

**MachineUtilization**
not available

**ComputationLoad**
not available because the graph surfing code is not operational at scale yet

**RunningTime**
[boisvert@cetuslac1 spate]$ grep TIMER test-spate-627-2048x16-1-MPI_nonblk.output
TIMER [Load input / Count input data] 36.146912 seconds
TIMER [Load input / Distribute input data] 2 minutes, 50.586319 seconds
TIMER [Load input] 3 minutes, 26.733231 seconds
TIMER [Build assembly graph / Distribute vertices] 6 minutes, 19.435547 seconds
TIMER [Build assembly graph / Distribute arcs] 7 minutes, 26.814911 seconds
TIMER [Build assembly graph] 13 minutes, 46.250427 seconds

**MemoryUtilization**
**Checksum**
[boisvert@cetuslac1 spate]$ sha1sum test-spate-627-2048x16-1-MPI_nonblk/coverage_distribution.txt-canonical 
01a293db48518190038eaddbaed8a47ca0323fc7  test-spate-627-2048x16-1-MPI_nonblk/coverage_distribution.txt-canonical

**GoodComments**
**BadComments**
graph traversal is not ready according to this job.

**NeutralComments**
